contents/c++/rvalue.md

```diff
@@ -0,0 +1,169 @@
+
+One example implementation of a move constructor and move assignment is shown below:
+```cpp
+Foo(Foo&& other) { // move constructor
+    x = other.x;
+    y = other.y;
+    z = other.z;
+}
+
+Foo& operator=(Foo&& other) { // move assignment
+    x = std::move(other.x);
+    y = std::move(other.y);
+    z = std::move(other.z);
+    return *this;
+}
+```
+
+In this move constructor and assignment, the contents of the `other` parameter is *moved* into the object, and the contents in `other` is destroyed afterwards. No additional memory allocation is required, and the *move* operation is done quickly by a few assignment of address operations, leading to a faster and more memory efficient program.
+
+Along with Move Semantics in `C++11`, the [STL library](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) provides the overloaded *move* functions for its container classes (e.g. `vector`, `list`, `set`), thus we can take advantage of these Move Semantics by simply supplying rvalues, without the need to redefine the classes ourselves.
+
+## rvalue anti-pattern
+After learning about the new rvalue references and Move Semantics in C++11, many programmers from the older C++ eras may fall into a trap of the rvalue [anti-pattern](https://en.wikipedia.org/wiki/Anti-pattern).
+
+Consider these 2 class constructors:
+```cpp
+Foo(const std::string& x, const std::string& y) { // copy constructor
+    _x = x;
+    _y = y;
+}
+
+Foo(std::string&& x, std::string&& y) { // move constructor
+    _x = x;
+    _y = y;
+}
+```
+
+As shown above, the *move* constructor defines each of its parameter to be rvalues. In this case, if the constructor is called using a mixture of both lvalues and rvalues, such as lvalue for `x` and rvalue for `y`, the copy constructor instead will be called! This is because the *copy* constructor is able to accept both lvalues and rvalues, while the *move* constructor is only able to accept rvalues. As such, we may think we have made use of Move Semantics to optimize our program, but that may not be the case!
+
+One solution to this problem would be to overload the constructor for each combination of rvalue parameters possible, like this:
+```cpp
+Foo(const std::string& x, const std::string& y) { // copy constructor
+    _x = x;
+    _y = y;
+}
+
+Foo(std::string&& x, const std::string& y) { // move constructor 1
+    _x = x;
+    _y = y;
+}
+
+Foo(const std::string& x, std::string&& y) { // move constructor 2
+    _x = x;
+    _y = y;
+}
+
+Foo(std::string&& x, std::string&& y) { // move constructor 3
+    _x = x;
+    _y = y;
+}
+```
+However, this is not feasible as we would require `2^n` overload functions, where `n` is the number of parameters. This results in large amount of [boilerplate code](https://www.quora.com/What-is-boilerplate-code), which in turn reduces code quality, and also contributes to increased memory consumption and compilation time.
```

Comment:
results in **increased** boilerplate code, which in turn reduces code quality, and ~~also contributes to~~ increases memory consumption and compilation time.